redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,639 @@
+/// This module contains the implementation of scanning operations in a Redis cluster.
+///
+/// The `ClusterScanArgs` struct represents the arguments for a cluster scan operation,
+/// including the scan state reference, match pattern, count, and object type.
+///
+/// The `ScanStateWrapper` struct is a wrapper for managing the state of a scan operation in a cluster.
+/// It holds a reference to the scan state and provides methods for accessing the state.
+///
+/// The `ClusterInScan` trait defines the methods for interacting with a Redis cluster during scanning,
+/// including retrieving node information, refreshing slot mapping, and routing commands to specific nodes.
+///
+/// The `ScanState` struct represents the state of a scan operation in a Redis cluster.
+/// It holds information about the current scan state, including the cursor position, scanned slots map,
+/// slots in the scan map, node being scanned, and node's epoch.
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use crate::aio::ConnectionLike;
+use crate::cluster_async::{ClusterConnInner, Connect, InnerCore};
+use crate::cluster_routing::SlotAddr;
+use crate::{cmd, from_redis_value, Cmd, ErrorKind, RedisError, RedisResult, Value};
+
+const BITS_PER_U64: usize = 64;
+const NUM_OF_SLOTS: usize = 16384;
+const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
+const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct ScanState {
+    pub cursor: u64,
+    pub scanned_slots_map: SlotsBitsArray,
+    pub slots_in_scan_map: SlotsBitsArray,
+    pub node_in_scan: String,
+    pub node_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    scan_state_ref: ScanStateWrapper,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<String>,
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_ref: ScanStateWrapper,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<String>,
+    ) -> Self {
+        Self {
+            scan_state_ref,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateWrapper {
```

Comment:
I dont think it will less confusing, because it make them think that they actually having the state and maybe can manipulate it. What about ScanStateRef?

Reply:
They want be confused that they can manipulate it if we won't provide any functions for that

Reply:
I dont like the idea of calling it ScanState, it gives the feeling that its somehow mutable or available and mean something to the user.
I'll go with ScanStateCursor.